network/newstream/cursors_test.go

```diff
@@ -0,0 +1,646 @@
+		if err = sim.Net.Connect(id, pivotEnode); err != nil {
+			t.Fatal(err)
+		}
+		nodeCount++
+		// wait for node to be set
+		time.Sleep(100 * time.Millisecond)
+	}
+	t.Fatal("node with po<=depth not found")
+	return
+}
+
+// waitForCursors checks if the pivot node has some cursors or not
+// by periodically checking for them.
+func waitForCursors(t *testing.T, sim *simulation.Simulation, pivotEnode, lookupEnode enode.ID, wantSome bool) {
+	t.Helper()
+
+	var got int
+	for i := 0; i < 1000; i++ { // 10s total wait
+		time.Sleep(10 * time.Millisecond)
+		s, ok := sim.Service(serviceNameSlipStream, pivotEnode).(*SlipStream)
+		if !ok {
+			continue
+		}
+		p := s.getPeer(lookupEnode)
+		if p == nil {
+			continue
+		}
+		got = len(p.getCursorsCopy())
+		if got != 0 == wantSome {
+			return
+		}
+	}
+	if wantSome {
+		t.Fatalf("got %v cursors, but want some", got)
+	} else {
+		t.Fatalf("got %v cursors, but want none", got)
+	}
+}
+
+// compareNodeBinsToStreams checks that the values on `onesCursors` correlate to the values in `othersBins`
+// onesCursors represents the stream cursors that node A knows about node B (i.e. they shoud reflect directly in this case
+// the values which node B retrieved from its local store)
+// othersBins is the array of bin indexes on node B's local store as they were inserted into the store
+func compareNodeBinsToStreams(t *testing.T, onesCursors map[string]uint64, othersBins []uint64) (err error) {
+	if len(onesCursors) == 0 {
+		return errors.New("no cursors")
+	}
+	if len(othersBins) == 0 {
+		return errors.New("no bins")
+	}
+
+	for nameKey, cur := range onesCursors {
+		id, err := strconv.Atoi(parseID(nameKey).Key)
+		if err != nil {
+			return err
+		}
+		if othersBins[id] != uint64(cur) {
+			return fmt.Errorf("bin indexes not equal. bin %d, got %d, want %d", id, cur, othersBins[id])
+		}
+	}
+	return nil
+}
+
+func compareNodeBinsToStreamsWithDepth(t *testing.T, onesCursors map[string]uint64, othersBins []uint64, depth uint) (err error) {
+	log.Debug("compareNodeBinsToStreamsWithDepth", "cursors", onesCursors, "othersBins", othersBins, "depth", depth)
+	if len(onesCursors) == 0 || len(othersBins) == 0 {
+		return errors.New("no cursors")
+	}
+	// inclusive test
+	for nameKey, cur := range onesCursors {
+		bin, err := strconv.Atoi(parseID(nameKey).Key)
+		if err != nil {
+			return err
+		}
+		if uint(bin) < depth {
+			return fmt.Errorf("cursor at bin %d should not exist. depth %d", bin, depth)
+		}
+		if othersBins[bin] != uint64(cur) {
+			return fmt.Errorf("bin indexes not equal. bin %d, got %d, want %d", bin, cur, othersBins[bin])
+		}
+	}
+
+	// exclusive test
+	for i := 0; i < int(depth); i++ {
+		// should not have anything shallower than depth
+		id := NewID("SYNC", fmt.Sprintf("%d", i))
+		if _, ok := onesCursors[id.String()]; ok {
+			return fmt.Errorf("oneCursors contains id %s, but it should not", id)
+		}
+	}
+	return nil
+}
+
+func TestCorrectCursorsExchangeRace(t *testing.T) {
```

Comment:
The mental taxation in understanding what this is doing is a bit too high for my taste. I'm happy we have started to push away complexity from production Swarm code, but we should probably rethink the complexity in tests or at least be mindful about that.

It is not immediately obvious what this test is testing. Tests should generally be simple and I shouldn't have to read the code and think for 15min to figure that out. What is this testing?

1. We are using a simulation framework.
2. We are mocking a devp2p protocol and injecting some custom hook
3. We are building peer fixtures using the p2p library.

We should probably be building more tools to write tests in a more simple manner.

No action required (or at least I don't know what I would expect), but this is def. not the direction we should be going in general IMO.